src/tree.rs

```diff
@@ -135,3 +149,125 @@ pub fn minimum_spanning_tree(
 
     Ok(spanning_tree)
 }
+
+/// Bipartition tree by finding balanced cut edges of a spanning tree using
+/// node contraction. Assumes that the tree is connected and is a spanning tree.
+/// A balanced edge is defined as an edge that, when cut, will split the
+/// population of the tree into two connected subtrees that have population near
+/// the population target within some epsilon. The function returns a list of
+/// all such possible cuts, represented as the set of nodes in one
+/// partition/subtree.
+///
+/// :param PyGraph graph: Spanning tree. Must be fully connected
+/// :param pops: The populations assigned to each node in the graph.
+/// :param float pop_target: The population target to reach when partitioning the
+///     graph.
+/// :param float epsilon: The maximum percent deviation from the pop_target
+///     allowed while still being a valid balanced cut edge.
+///
+/// :returns: A list of tuples, with each tuple representing a distinct
+///     balanced edge that can be cut. The tuple contains the root of one of the
+///     two partitioned subtrees and the set of nodes making up that subtree.
+#[pyfunction]
+#[pyo3(text_signature = "(spanning_tree, pop, target_pop, epsilon)")]
+pub fn bipartition_tree(
+    _py: Python,
+    spanning_tree: &graph::PyGraph,
+    pops: Vec<f64>,
+    pop_target: f64,
+    epsilon: f64,
+) -> PyResult<Vec<(usize, Vec<usize>)>> {
+    let mut pops = pops;
+    let spanning_tree_graph = &spanning_tree.graph;
+    let mut same_partition_tracker: Vec<Vec<usize>> =
+        vec![vec![]; spanning_tree_graph.node_count()]; // Keeps track of all all the nodes on the same side of the partition
```

Comment:
The other thing I just realized is how do you handle holes in the node indices here? If a node is removed that index will be missing from `graph.node_indices()` with `PyGraph`. For example if you do:

```python
import retworkx
g = retworkx.PyGraph()
g.add_nodes_from(range(5))
g.remove_node(2)
print(g.node_indices())
```

will return `[0, 1, 3, 4]`.

I think you need to adjust this to be `node_bound()` which is the `max(index)` in the node indices list instead of the number of nodes in the graph. There might be other adjustments you need to do to account for this.

Reply:
Ok, thanks for pointing this out -- just changed it to be `node_bound()`.